golang/http.go

```diff
@@ -0,0 +1,358 @@
+
+	_, err := h.processRequest(
+		`UPDATE guestbook.posts SET text=$1 WHERE id=$2`,
+		id,
+		req.Text,
+	)
+	if err != nil {
+		return http.StatusInternalServerError, nil, fmt.Errorf("can't update posts table: %s", err)
+	}
+
+	if err := h.refreshTable("guestbook.posts"); err != nil {
+		return http.StatusInternalServerError, nil, err
+	}
+
+	details, err := h.getPostDetails(id)
+	if err != nil {
+		return http.StatusInternalServerError, nil, fmt.Errorf("can't retrieve added post: %s", err)
+	}
+
+	return http.StatusOK, details, nil
+}
+
+func (h Http) DeletePost(id string) (int, error) {
+
+	com, err := h.processRequest(
+		`DELETE FROM guestbook.posts WHERE id=$1`,
+		id,
+	)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("can't delete posts table: %s", err)
+	}
+
+	if com.RowCount == 0 {
+		return http.StatusNotFound, fmt.Errorf(`post with id=%s" not found`, id)
+
+	}
+
+	return http.StatusNoContent, nil
+}
+
+func (h Http) LikePost(id string) (int, *PostDetails, error) {
+
+	com, err := h.processRequest(
+		`UPDATE guestbook.posts SET like_count = like_count + 1 WHERE id=$1`,
+		id)
+	if err != nil {
+		return http.StatusInternalServerError, nil, fmt.Errorf(fmt.Sprintf("Can't increase likes count: %s", err))
+	}
+
+	if com.RowCount == 0 {
+		return http.StatusNotFound, nil, fmt.Errorf(`post with id=%s" not found`, id)
+	}
+
+	details, err := h.getPostDetails(id)
+	if err != nil {
+		return http.StatusInternalServerError, nil, fmt.Errorf("can't retrieve post: %s", err)
+	}
+
+	return http.StatusOK, details[0], nil
+}
+
+func (h Http) GetAllImages() (int, []*AllImagesResponse, error) {
+
+	response, err := h.processRequest(`SELECT digest, last_modified FROM blob.guestbook_images`)
+	if err != nil {
+		return http.StatusInternalServerError, nil, err
+	}
+
+	result, err := processResponse[AllImagesResponse](response)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	return http.StatusOK, result, nil
+}
+
+func processResponse[T any](resp *CrateResponse) ([]*T, error) {
+
+	result := []*T{}
+
+	for _, row := range resp.Rows {
+		if row == nil {
+			continue
+		}
+
+		// preparing map
+		m := make(map[string]interface{})
+		for i, col := range resp.Cols {
+			m[col] = row[i]
+		}
+
+		// hook for converting of timestamps to time.Time
+		toTimeHookFunc := func() mapstructure.DecodeHookFunc {
+			return func(
+				f reflect.Type,
+				t reflect.Type,
+				data interface{}) (interface{}, error) {
+				if t != reflect.TypeOf(time.Time{}) {
+					return data, nil
+				}
+
+				switch f.Kind() {
+				case reflect.String:
+					return time.Parse(time.RFC3339, data.(string))
+				case reflect.Float64:
+					return time.Unix(0, int64(data.(float64))*int64(time.Millisecond)), nil
+				case reflect.Int64:
+					return time.Unix(0, data.(int64)*int64(time.Millisecond)), nil
+				default:
+					return data, nil
+				}
+				// Convert it by parsing
+			}
+		}
```

Comment:
Does this callback function has to be defined _inside_ the loop? Is it possible to defined it outside, to improve readability, and maybe also efficiency?

Reply:
applied